### DISCUSSION TITLE
组件开放健康检查接口

### DISCUSSION THREAD

**Discussion by mariodu** [midwayjs/midway#3354] [Ideas]
组件不健康的可能导致应用异常，需要用健康检查机制反馈，让应用摘除流量，防止异常情况发生。

```
外部 health check -> midway framework health check -> componet health check
```

**Comment by czy88840616** [comment]
技术方案： #3370

**Comment by flyingcrp** [comment]
有这个后可以知道项目依赖访问的所有组件或服务是否正常吗?

**Reply by czy88840616** [reply]
目前还是不知道的，需要组件去额外编写接入逻辑。

**Reply by flyingcrp** [reply]
升级一下组件的开发说明和把这个加进去可以. 但是老的组件也需要升级

**Reply by czy88840616** [reply]
部分组件需要，特别是 redis，数据库连接之类的。

**Reply by flyingcrp** [reply]
可以整理一下优先支持,是不是顺带会把服务发现也升级了 😆 

redis,数据库,队列,服务发现

**Reply by czy88840616** [reply]
服务发现是另一个事情。。。

**Reply by czy88840616** [reply]
简单写了 redis， #3473，想到几个问题

* 1、检查的行为，如果组件里有多个 client，那么是只发现第一个错误就返回，还是等都查询完再返回，比如业务有强弱依赖，是不是考虑如果是弱依赖，则健康检查依旧保持正常
* 2、返回的内容，目前单个组件只有一个 reson，是否有自定义的需求？如果用户想返回所有的状态

**Reply by flyingcrp** [reply]
1. 强弱依赖由组件自己定义，健康与否也应该由组件决定，框架只需要一个健康或不健康的结果。
2. 健康检查状态只有健康和非健康，不应该存在其他状态。这样更清晰更简单。
感觉组件内部多状态是不是不应该暴露出来。组件只要有自己的健康标准就行。

**Reply by czy88840616** [reply]
最开始也是这么想的，不过现在的确健康检查的状态是会影响到业务承接流量的，如果一个弱依赖有问题，那么健康检查都标记有问题进而造成所有容器流量摘除，这个设计感觉有点过了。

另一方面组件都是通用的，无法知道哪个 client 是弱依赖，这就导致了无法分级去处理这个行为，不过可以在配置 client 的时候加上强弱依赖标识，标准化一下，进而做处理也行。
